### PR TITLE
Treatment Strategy and Individual Outcome Estimation implementation.

### PR DESCRIPTION
**Pull request by JulinaM**
Pull Request for the commits related with **Individual Outcome Estimation.**

### CODE REVIEW

File: causallib/time_varying/treament_strategy.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+from abc import abstractmethod
+
+
+class TreatmentStrategy:
+    def __init__(self, source_t, **kwargs):
```
Comment (pchakraborty1) [inline causallib/time_varying/treament_strategy.py]: `source_t` should not be in init

File: causallib/time_varying/treament_strategy.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        self.source_t = source_t
+
+    def __call__(self):
```
Comment (pchakraborty1) [inline causallib/time_varying/treament_strategy.py]: arguments should come in here

File: causallib/time_varying/treament_strategy.py (hunk 1 above)
Comment (pchakraborty1) [reply causallib/time_varying/treament_strategy.py]: Change name of `source_t` to `history_X`, `history_a`, `history_y`

--> essentially make the api consistent with how we are defining the counterfactual simulator

File: causallib/time_varying/treament_strategy.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        self.get_action()
+
+    @abstractmethod
+    def get_action(self):
```
Comment (pchakraborty1) [inline causallib/time_varying/treament_strategy.py]: For abstract class, if you think there will be other parameters apart from the histories, pass these as `**kwargs`

File: causallib/time_varying/treament_strategy.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        # return out.unsqueeze(1)
+
+
+class CFBernoulli(TreatmentStrategy):
+    """
+           CounterFactual Bernoulli Treatment Strategy
+    """
+    def __init__(self,
+                 source_t,
```
Comment (pchakraborty1) [inline causallib/time_varying/treament_strategy.py]: parameterize this with the probability `p`

i.e. an always treat strategy would be `p=1.`, a never treat would be `p=0.` and then in general we can have a naive Bernoulli strategy with `p \in [0, 1]`

File: causallib/time_varying/treament_strategy.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        """
+        raise NotImplementedError
+
+
+class Observational(TreatmentStrategy):
+    """
+        Observational Treatment Strategy
+    """
+    def __init__(self,
+                 source_t,
```
Comment (pchakraborty1) [inline causallib/time_varying/treament_strategy.py]: Remove `source_t` from here and add it to `call`

File: causallib/time_varying/GFormula.py [hunk 6]
```diff
@@ -4,12 +4,23 @@
...
         GFormula class that is based on Monte Carlo Simulation for creating the noise.
     """
+    def __init__(self, outcome_model, treatment_model, covariate_models, refit_models, seed, n_obsv, n_sims, n_steps, mode, resid_val):
+        super(GFormula, self).__init__(outcome_model, treatment_model, covariate_models, refit_models)
+        self.seed = seed
+        self.n_obsv = n_obsv
+        self.n_sims = n_sims
+        self.n_steps = n_steps
+        self.mode = mode
+        self.resid_val = resid_val
```
Comment (pchakraborty1) [inline causallib/time_varying/GFormula.py:23]: If models are fitted via the `GFormula` then it is calculated. 

so one suggestions, don't put this in init. have a separate setter method, which fit can use or externally, people can provide

File: causallib/time_varying/GFormula.py [hunk 7]
```diff
@@ -4,12 +4,23 @@
...
 class GFormula(GMethodBase):
     """
         GFormula class that is based on Monte Carlo Simulation for creating the noise.
     """
+    def __init__(self, outcome_model, treatment_model, covariate_models, refit_models, seed, n_obsv, n_sims, n_steps, mode, resid_val):
+        super(GFormula, self).__init__(outcome_model, treatment_model, covariate_models, refit_models)
+        self.seed = seed
+        self.n_obsv = n_obsv
+        self.n_sims = n_sims
+        self.n_steps = n_steps
```
Comment (pchakraborty1) [inline causallib/time_varying/GFormula.py:21]: move it to estimate individual outcome. perhaps derive it from the existing params

File: causallib/time_varying/GFormula.py [hunk 8]
```diff
@@ -4,12 +4,23 @@
...
+import torch as T #TODO remove torch dependency
 
 
 class GFormula(GMethodBase):
     """
         GFormula class that is based on Monte Carlo Simulation for creating the noise.
     """
+    def __init__(self, outcome_model, treatment_model, covariate_models, refit_models, seed, n_obsv, n_sims, n_steps, mode, resid_val):
+        super(GFormula, self).__init__(outcome_model, treatment_model, covariate_models, refit_models)
+        self.seed = seed
```
Comment (pchakraborty1) [inline causallib/time_varying/GFormula.py:18]: in scikit-learn, the parameter is `random_state` which ties back to numpy. we can perhaps use that

From sklearn documentation: https://scikit-learn.org/dev/glossary.html#term-random_state

File: causallib/time_varying/GFormula.py [hunk 9]
```diff
@@ -46,22 +57,101 @@ def estimate_individual_outcome(self,
                                     a: pd.Series, 
                                     t: Optional[pd.Series] = None,
                                     y: Optional[Any] = None,
-                                    treatment_strategy: Callable = None,
```
Comment (pchakraborty1) [inline causallib/time_varying/GFormula.py:49]: bring back the `treatment_strategy` -> we would need one to simulate

File: causallib/time_varying/base.py [hunk 10]
```diff
@@ -44,12 +44,14 @@ class GMethodBase(TimeVaryingBaseEstimator):
         GFormula base Estimator
     """
     def __init__(self,
-                 treatment_model: Any,
+                 treatment_strategy: Callable,
```
Comment (pchakraborty1) [inline causallib/time_varying/base.py]: needs to be reverted back

File: causallib/time_varying/base.py [hunk 11]
```diff
@@ -133,7 +131,6 @@ def estimate_population_outcome(self,
                                     a: pd.Series,
                                     t: pd.Series,
                                     y: Optional[Any] = None,
-                                    treatment_strategy: Callable = None,
```
Comment (pchakraborty1) [inline causallib/time_varying/base.py]: bring this back

File: causallib/time_varying/treament_strategy.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+        """
+            inverse_transform (Callable): Scaler class that compute the means and std to be used for later scaling.
+                                        eg: sklearn.StandardScaler()
+            kwargs (dict): Optional kwargs for init call in TreatmentStrategy
+        """
+        super(TreatmentStrategy, self).__init__(**kwargs)
+        self.inverse_transform = inverse_transform
+
+    def get_action(self,
+                   prev_x,
```
Comment (pchakraborty1) [inline causallib/time_varying/treament_strategy.py:34]: change variable names to match the API

File: causallib/time_varying/GFormula.py (hunk 6 above)
Comment (pchakraborty1) [reply causallib/time_varying/GFormula.py:23]: @JulinaM can you see this?

File: causallib/time_varying/GFormula.py [hunk 13]
```diff
@@ -46,29 +58,108 @@ def estimate_individual_outcome(self,
...
+                if t <= n_margin: # and debug:  # == 11:
+                    # print(T.allclose(hidden[0], hidden_p[0]))
+                    # print(T.allclose(hidden[1], hidden_p[1]))
+                    print(T.cat(simulation['prediction'], dim=1).squeeze())
+                    print(T.cat(simulation['covariates'], dim=1).squeeze())
+
+            simulation['actions'] = T.cat(simulation['actions'], dim=1)  # N_sim * n_steps * F-act
+            simulation['prediction'] = T.cat(simulation['prediction'], dim=1)
+            simulation['covariates'] = T.cat(simulation['covariates'], dim=1)  # N_sim * n_steps * F-act
+        # return res
```
Comment (JulinaM) [inline causallib/time_varying/GFormula.py]: @ehudkr, I have a confusion here. What should be the return type in this method? As per the framework, the return type is a data frame for predicting only treatments. But here we also have predicted covariates along with treatment values. So could you suggest us how should we move ahead with the return types?

**Comment by JulinaM** [comment]
Technical debt remaining in `causallib/time_varying/GFormula.py`:

1. `estimate_individual_outcome` 
    - add up more details in docs 
    - add columns names and sample_id as a column before returning df

2. `_estimate_individual_outcome_single_sample`
     - prepare input to support for RNN later 

3. `estimate_population_outcome`
     - logic needs to be refined

4. `_apply_noise`
     - convert Torch to np
     - infer box_type from data

5. `_prepare_data`
     -  write actual code for data manipulation

6. `_init_models`
     -  write actual code for model initialization
     -   initially do it for only sklearn

7. `_predict`
     -  convert torch to np
     -  debug with actual sklearn model and data

**Review by pchakraborty1 — APPROVED** [review]
Remaining comments will be used as technical debt

File: causallib/time_varying/treament_strategy.py (hunk 1 above)
Comment (JulinaM) [reply causallib/time_varying/treament_strategy.py]: The changes have already been made in [f31951a4f22955ce8f017e56d1909ae38c06188c](https://github.com/pchakraborty1/causallib/commit/f31951a4f22955ce8f017e56d1909ae38c06188c) commit .

File: causallib/time_varying/treament_strategy.py (hunk 2 above)
Comment (JulinaM) [reply causallib/time_varying/treament_strategy.py]: The changes have already been made in [7d114d1aeafc95830d436d5667409df7d1d83c27](https://github.com/pchakraborty1/causallib/commit/7d114d1aeafc95830d436d5667409df7d1d83c27) commit.

File: causallib/time_varying/treament_strategy.py (hunk 4 above)
Comment (JulinaM) [reply causallib/time_varying/treament_strategy.py]: The changes have already been made in [f31951a4f22955ce8f017e56d1909ae38c06188c](https://github.com/pchakraborty1/causallib/commit/f31951a4f22955ce8f017e56d1909ae38c06188c) commit .

File: causallib/time_varying/treament_strategy.py (hunk 5 above)
Comment (JulinaM) [reply causallib/time_varying/treament_strategy.py]: The changes have already been made in [f31951a4f22955ce8f017e56d1909ae38c06188c](https://github.com/pchakraborty1/causallib/commit/f31951a4f22955ce8f017e56d1909ae38c06188c) commit.

File: causallib/time_varying/base.py (hunk 10 above)
Comment (JulinaM) [reply causallib/time_varying/base.py]: Already reverted in [336f405529db8f9c0c9e9f726e8eedb44567a8f9](https://github.com/pchakraborty1/causallib/commit/336f405529db8f9c0c9e9f726e8eedb44567a8f9) commit.

File: causallib/time_varying/treament_strategy.py (hunk 3 above)
Comment (JulinaM) [reply causallib/time_varying/treament_strategy.py]: No extra args at the moment.

File: causallib/time_varying/GFormula.py (hunk 9 above)
Comment (JulinaM) [reply causallib/time_varying/GFormula.py:49]: Already reverted in [336f405529db8f9c0c9e9f726e8eedb44567a8f9](https://github.com/pchakraborty1/causallib/commit/336f405529db8f9c0c9e9f726e8eedb44567a8f9) commit.

File: causallib/time_varying/treament_strategy.py (hunk 12 above)
Comment (JulinaM) [reply causallib/time_varying/treament_strategy.py:34]: for now, we are naming "prev_x". We will change it to "history_x" in later version.

**Pull request merged**